### PR TITLE
Let sysrepo-plugind write a pid file

### PR DESCRIPTION
**Pull request by jorgensigvardsson**
In systems where daemons depend on each other, it is useful to leave pid
files under /var/run so that init (PID 1) can track dependencies between
daemons, and thus orchestrate their startup times.

The implementation is inspired by the implementation found in netopeer2.

Signed-off-by: Jörgen Sigvardsson <jorgen.sigvardsson@gmail.com>

### CODE REVIEW

**Comment by jorgensigvardsson** [comment]
I had worked against an older version of the master branch, hence I got conflicts. When I rebased my branch onto an updated master, I could see that similar changes had been made for systemd. Should I take systemd into account? If running systemd, you probably don't want pid files left and right? Let me know, and I'll adjust for systemd.

File: src/executables/sysrepo-plugind.c
Comment (jktjkt) [inline src/executables/sysrepo-plugind.c:555]: This change will effectively make it impossible to run `sysrepo-plugind` as non-`root` by default. Since `/var/run` is not writable by all, unless people take an explicit action to build with an option to put the PID file into a writable location, `sysrepo-plugind` with no arguments will fail to start on a developer's box.

Instead of a build-time option for a default PID file location, and then a runtime option for overriding that, why not just do nothing by default, *and* allow the user to pass a flag like `--pid-file=/var/run/sysrepo-plugind.pid` which will activate this mechanism?

File: src/executables/sysrepo-plugind.c
Comment (jktjkt) [inline src/executables/sysrepo-plugind.c:490]: There's no matching call to `free(pidfile)`, so this leaks memory.

File: src/executables/sysrepo-plugind.c
Comment (jktjkt) [inline src/executables/sysrepo-plugind.c:442]: This is fragile. On Linux, the upper bound on the maximal PID values [has been increased before](https://lwn.net/Articles/10424/), and the fact that it's currently limited to 4M, and therefore will *currently* fit into seven digits plus the trailing `NULL` is just an implementation detail.

**Review by jktjkt — CHANGES_REQUESTED** [review]
Hi, the development happens in the `devel` branch, so PRs should go there.

Since I'm personally very happy with `systemd`, I think that implementing this daemonization and PID-file-writing code in each and every daemon is a bit tedious, but that's probably my personal bias -- and this binary will already daemonize by default anyway.

That said, I think that the code will be simpler and more straightforward if the PID file writing is explicitly activated by a runtime option -- with no default location.

File: src/executables/sysrepo-plugind.c
Comment (jorgensigvardsson) [reply src/executables/sysrepo-plugind.c:609]: Thanks for spotting that one!

File: src/executables/sysrepo-plugind.c
Comment (jorgensigvardsson) [reply src/executables/sysrepo-plugind.c:555]: Great idea, thanks!

File: src/executables/sysrepo-plugind.c
Comment (jorgensigvardsson) [reply src/executables/sysrepo-plugind.c:490]: I was going to rely on the OS to reclaim memory, but I can see memory checking tools report it.

**Comment by jorgensigvardsson** [comment]
Should I make the pid file a compile time option, just like the systemd integration? Either way works for me (I'm building from source).

Also, I'll rebase onto the devel branch.

By the way, do you know the state of all components involved (to get netopeer2, sysrepo and libyang) on the devel branch(s)? I tried verify my changes by building from master on each, but got some bad error messages about contexts being invalid and whatnot. It could've been me though, because it was Friday evening after a long week. 🥴  Ended up having to comment out irrelevant code paths in srpd and test just the pidfile code path.

Thanks for the reviewn

**Comment by jktjkt** [comment]
> Should I make the pid file a compile time option, just like the systemd integration? Either way works for me (I'm building from source).

Better wait for @michalvasko's reply, he's the maintainer.

> By the way, do you know the state of all components involved (to get netopeer2, sysrepo and libyang) on the devel branch(s)? I tried verify my changes by building from master on each, but got some bad error messages about contexts being invalid and whatnot. It could've been me though, because it was Friday evening after a long week. woozy_face Ended up having to comment out irrelevant code paths in srpd and test just the pidfile code path.

Our team tends to build from `devel` of everything quite regularly, so I don't really care that much about `master` or about releases, but I think the maintainers do care :). In general, it seems to me that `master` is typically only updated prior to a release, so it "should" be stable. In the `devel` there's an occasional breakage, especially when more intrusive changes are merged -- and the NACM refactoring which went in this Friday looks like a good example of a possible breaking change.

File: src/executables/sysrepo-plugind.c
Comment (jktjkt) [reply src/executables/sysrepo-plugind.c:442]: It looks like I forgot to mention that I suggest to use `asprintf()` since it's something that's already used throughout the code.

**Comment by jorgensigvardsson** [comment]
Sorry, had to reopen a new pull request: https://github.com/sysrepo/sysrepo/pull/2745

**Pull request closed without merging**
